### PR TITLE
Correct decoding of foreign expansions during incr. comp.

### LINKED ISSUE #74946: Incremental recompilation bug while compiling sysroot
### Code

Assuming that you copied `$(rustc --print sysroot)/lib/rustlib/src/rust` to `./sysroot_src`.

```toml
#- Cargo.toml
[package]
authors = ["bjorn3 <bjorn3@users.noreply.github.com>"]
name = "sysroot"
version = "0.0.0"

[dependencies]
core = { path = "./sysroot_src/library/core" }
compiler_builtins = "0.1"
alloc = { path = "./sysroot_src/library/alloc" }
std = { path = "./sysroot_src/library/std", features = ["panic_unwind"] }

[patch.crates-io]
rustc-std-workspace-core = { path = "./sysroot_src/library/rustc-std-workspace-core" }
rustc-std-workspace-alloc = { path = "./sysroot_src/library/rustc-std-workspace-alloc" }
```

```rust
//- src/lib.rs
// empty
```

```sh
RUSTFLAGS="-Zmir-opt-level=2" cargo build; rm -r target/debug/{build,deps,examples}; RUSTFLAGS="-Zmir-opt-level=2" cargo build
```

### Meta

`rustc --version --verbose`:
```
rustc 1.47.0-nightly (db0492ace 2020-07-29)
binary: rustc
commit-hash: db0492ace429cfeb3567e2c04e300be7df9972ff
commit-date: 2020-07-29
host: x86_64-unknown-linux-gnu
release: 1.47.0-nightly
LLVM version: 10.0
```

### Error output

```
thread 'rustc' panicked at 'called `Option::unwrap()` on a `None` value', src/librustc_metadata/rmeta/decoder.rs:1886:22
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/libunwind.rs:86
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at library/std/src/sys_common/backtrace.rs:78
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at library/std/src/sys_common/backtrace.rs:59
   4: core::fmt::write
             at library/core/src/fmt/mod.rs:1117
   5: std::io::Write::write_fmt
             at library/std/src/io/mod.rs:1510
   6: std::sys_common::backtrace::_print
             at library/std/src/sys_common/backtrace.rs:62
   7: std::sys_common::backtrace::print
             at library/std/src/sys_common/backtrace.rs:49
   8: std::panicking::default_hook::{{closure}}
             at library/std/src/panicking.rs:198
   9: std::panicking::default_hook
             at library/std/src/panicking.rs:217
  10: rustc_driver::report_ice
  11: std::panicking::rust_panic_with_hook
             at library/std/src/panicking.rs:530
  12: rust_begin_unwind
             at library/std/src/panicking.rs:437
  13: core::panicking::panic_fmt
             at library/core/src/panicking.rs:85
  14: core::panicking::panic
             at library/core/src/panicking.rs:50
  15: rustc_span::hygiene::decode_expn_id
  16: <rustc_span::hygiene::SyntaxContextData as rustc_serialize::serialize::Decodable>::decode
  17: rustc_span::hygiene::decode_syntax_context
  18: <rustc_metadata::rmeta::decoder::DecodeContext as rustc_serialize::serialize::SpecializedDecoder<rustc_span::span_encoding::Span>>::specialized_decode
  19: <rustc_middle::mir::SourceInfo as rustc_serialize::serialize::Decodable>::decode
  20: <rustc_middle::mir::Statement as rustc_serialize::serialize::Decodable>::decode
  21: rustc_serialize::serialize::Decoder::read_seq
  22: <rustc_middle::mir::BasicBlockData as rustc_serialize::serialize::Decodable>::decode
  23: rustc_serialize::serialize::Decoder::read_seq
  24: <rustc_middle::mir::Body as rustc_serialize::serialize::Decodable>::decode
  25: rustc_metadata::rmeta::decoder::<impl rustc_metadata::creader::CrateMetadataRef>::get_optimized_mir
  26: rustc_metadata::rmeta::decoder::cstore_impl::provide_extern::optimized_mir
  27: rustc_middle::dep_graph::<impl rustc_query_system::dep_graph::DepKind for rustc_middle::dep_graph::dep_node::DepKind>::with_deps
  28: rustc_query_system::query::plumbing::load_from_disk_and_cache_in_memory
  29: rustc_query_system::query::plumbing::get_query_impl
  30: rustc_middle::ty::<impl rustc_middle::ty::context::TyCtxt>::instance_mir
  31: rustc_mir::monomorphize::collector::collect_neighbours
  32: rustc_mir::monomorphize::collector::collect_items_rec
  33: rustc_mir::monomorphize::collector::collect_items_rec
  34: rustc_mir::monomorphize::collector::collect_items_rec
  35: rustc_mir::monomorphize::collector::collect_items_rec
  36: rustc_mir::monomorphize::collector::collect_items_rec
  37: rustc_mir::monomorphize::collector::collect_items_rec
  38: rustc_mir::monomorphize::collector::collect_items_rec
  39: rustc_mir::monomorphize::collector::collect_items_rec
  40: rustc_mir::monomorphize::collector::collect_items_rec
  41: rustc_mir::monomorphize::collector::collect_items_rec
  42: rustc_mir::monomorphize::collector::collect_items_rec
  43: rustc_mir::monomorphize::collector::collect_items_rec
  44: rustc_mir::monomorphize::collector::collect_items_rec
  45: rustc_mir::monomorphize::collector::collect_items_rec
  46: rustc_mir::monomorphize::collector::collect_items_rec
  47: rustc_mir::monomorphize::collector::collect_items_rec
  48: rustc_mir::monomorphize::collector::collect_items_rec
  49: rustc_mir::monomorphize::collector::collect_items_rec
  50: rustc_mir::monomorphize::collector::collect_items_rec
  51: rustc_mir::monomorphize::collector::collect_items_rec
  52: rustc_mir::monomorphize::collector::collect_items_rec
  53: rustc_mir::monomorphize::collector::collect_items_rec
  54: rustc_mir::monomorphize::collector::collect_items_rec
  55: rustc_session::utils::<impl rustc_session::session::Session>::time
  56: rustc_mir::monomorphize::collector::collect_crate_mono_items
  57: rustc_mir::monomorphize::partitioning::collect_and_partition_mono_items
  58: rustc_middle::ty::query::<impl rustc_query_system::query::config::QueryAccessors<rustc_middle::ty::context::TyCtxt> for rustc_middle::ty::query::queries::collect_and_partition_mono_items>::compute
  59: rustc_middle::dep_graph::<impl rustc_query_system::dep_graph::DepKind for rustc_middle::dep_graph::dep_node::DepKind>::with_deps
  60: rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl
  61: rustc_data_structures::stack::ensure_sufficient_stack
  62: rustc_query_system::query::plumbing::force_query_impl
  63: rustc_middle::ty::query::force_from_dep_node
  64: rustc_query_system::dep_graph::graph::DepGraph<K>::try_mark_previous_green
  65: rustc_query_system::dep_graph::graph::DepGraph<K>::try_mark_green_and_read
  66: rustc_data_structures::stack::ensure_sufficient_stack
  67: rustc_query_system::query::plumbing::get_query_impl
  68: rustc_metadata::rmeta::encoder::EncodeContext::encode_crate_root
  69: rustc_metadata::rmeta::encoder::encode_metadata_impl
  70: rustc_data_structures::sync::join
  71: rustc_metadata::rmeta::decoder::cstore_impl::<impl rustc_middle::middle::cstore::CrateStore for rustc_metadata::creader::CStore>::encode_metadata
  72: rustc_middle::ty::context::TyCtxt::encode_metadata
  73: rustc_interface::passes::start_codegen
  74: rustc_middle::ty::context::tls::enter_global
  75: rustc_interface::queries::Queries::ongoing_codegen
  76: rustc_interface::queries::<impl rustc_interface::interface::Compiler>::enter
  77: rustc_span::with_source_map
  78: rustc_interface::interface::create_compiler_and_run
  79: scoped_tls::ScopedKey<T>::set
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.47.0-nightly (db0492ace 2020-07-29) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z mir-opt-level=2 -C prefer-dynamic -C embed-bitcode=no -C debuginfo=2 -C incremental --crate-type dylib --crate-type rlib

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [optimized_mir] optimizing MIR for `<alloc_crate::vec::IntoIter<T> as core::iter::Iterator>::next`
#1 [collect_and_partition_mono_items] collect_and_partition_mono_items
#2 [exported_symbols] exported_symbols
end of query stack
```

</p>
</details>

### PR DESCRIPTION
**Pull request by cjgillot**
Fixes https://github.com/rust-lang/rust/issues/74946

The original issue was due to a wrong assertion in `expn_hash_to_expn_id`.

The secondary issue was due to a mismatch between the encoding and decoding paths for expansions that are created after the TyCtxt is created.

### CODE REVIEW

File: src/test/incremental/mir-opt.rs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+// MIR optimizations can create expansions after the TyCtxt has been created.
```
Comment (petrochenkov) [inline src/test/incremental/mir-opt.rs:1]: Is this about `ExpnKind::Inlined`?
It's a recent addition, and I never manage to get to investigating what it does.
Looks like a total misuse of the expansion machinery at first glance.

Comment (cjgillot) [reply src/test/incremental/mir-opt.rs:1]: Yes, `ExpnKind::Inlined` is involved in this case. I did not try to remove it: I find the possibility to create expansions after the TyCtxt exists interesting for (eventually) incremental macro expansion.

**Pull request merged**